plutus-chain-index/app/Indexer.hs

```diff
@@ -84,44 +90,119 @@ parseHash hash =
   deserialiseFromRawBytesHex (proxyToAsType Proxy) (encodeUtf8 $ pack hash)
 
 getDatums :: BlockInMode CardanoMode -> [(SlotNo, (DatumHash, Datum))]
-getDatums (BlockInMode (Block (BlockHeader slotNo _ _) txs) era) = withIsCardanoEra era $ concatMap (go era) txs
-  where
-    go :: C.IsCardanoEra era
-       => C.EraInMode era C.CardanoMode
-       -> C.Tx era
-       -> [(SlotNo, (DatumHash, Datum))]
-    go era' tx =
-      let hashes = either (const []) (assocs . _citxData) $ fromCardanoTx era' tx
+getDatums (BlockInMode (Block (BlockHeader slotNo _ _) txs) era) =
+  withIsCardanoEra era $ concatMap (getDatum slotNo era) txs
+
+getDatum
+  :: C.IsCardanoEra era
+  => SlotNo
+  -> C.EraInMode era C.CardanoMode
+  -> C.Tx era
+  -> [(SlotNo, (DatumHash, Datum))]
+getDatum slotNo era tx =
+      let hashes = either (const []) (assocs . _citxData) $ fromCardanoTx era tx
```

Comment:
I would avoid using `fromCardanoTx` here. It's defined in `plutus-chain-index-core` while both the inputs and the outputs of this function only need `plutus-ledger-api` (which is a dependency of `plutus-chain-index-core`).

Reply:
We won't need this once we use the streaming component right?

Reply:
It's not the streaming component that gets data out of the blocks, it's the stream consumer that does that.
That said, the current version on main doesn't use `fromCardanoTx` so perhaps this just needs a rebase
https://github.com/input-output-hk/plutus-apps/blob/main/plutus-chain-index/app/Marconi.hs#L103

Other functions below have been added though.

Reply:
I was thinking of this version

https://github.com/input-output-hk/plutus-apps/blob/main/plutus-streaming/examples/Example2.hs#L32-L38

Reply:
I think this is a good idea, especially as we should be trying to move away from chain-index specific data structures in order to eventually phase them out as marconi starts to get center-stage.